lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py

```diff
@@ -1,31 +1,167 @@
+        assert exitcode == 0
+
+    def test_non_existent_group(self, tool, agent_config):
+        command = ["pbench-list-tools", "--group", "unknown"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert exitcode == 1
+
+    def test_non_existent_name(self, tool, agent_config):
+        command = ["pbench-list-tools", "-n", "unknown"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        # XXX 0.69.9 backward compatibility: non-existent names do not return errors
+        assert exitcode == 0
+
+    def test_existing_group_name(self, tool, agent_config):
+        command = ["pbench-list-tools", "--group", "default", "--name", "perf"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert b"tool name: perf groups: default\n" == out
+        assert exitcode == 0
+
+    def test_existing_group_non_existent_name(self, tool, agent_config):
+        command = ["pbench-list-tools", "--group", "default", "--name", "unknown"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert exitcode == 0
+
+    def test_non_existent_group_existing_name(self, tool, agent_config):
+        command = ["pbench-list-tools", "--group", "unknown", "--name", "perf"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert exitcode == 1
+
+
+class Test_list_tool_tools_registered_with_options:
+    @pytest.fixture
+    def tool(self, pbench_run):
+        for group in ["default", "test"]:
+            p = pbench_run / f"tools-v1-{group}" / "testhost.example.com"
+            p.mkdir(parents=True)
+            tool = p / "iostat"
+            tool.write_text("--interval=30")
+            tool = p / "mpstat"
+            tool.write_text("--interval=300")
+        return
+
+    def test_existing_group_options(self, tool, agent_config):
+        command = ["pbench-list-tools", "--group", "default", "--with-option"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        # XXX this is the current output - pretty ugly
+        assert (
+            b"default: testhost.example.com [('iostat', '--interval=30'), ('mpstat', '--interval=300')]\n"
+            == out
+        )
+        assert exitcode == 0
+
+    def test_non_existent_group_options(self, tool, agent_config):
+        command = ["pbench-list-tools", "--group", "unknown", "--with-option"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert exitcode == 1
+
+    def test_existing_group_name_options(self, tool, agent_config):
+        command = [
+            "pbench-list-tools",
+            "--group",
+            "default",
+            "--name",
+            "mpstat",
+            "--with-option",
+        ]
+        out, err, exitcode = pytest.helpers.capture(command)
+        # XXX this is the current output - obviously wrong
+        assert b"tool name: mpstat groups: default\n" == out
+        assert exitcode == 0
```

Comment:
We should not be adding test cases which validate incorrect results -- this test should be marked for [skipping](https://docs.pytest.org/en/latest/how-to/skipping.html) until we have the tool producing the output that we desire.

Reply:
Thanks for that. I'll keep it in mind.

For now, I decided to try an implementation of  what seems to me a perfectly valid interpretation: list the tool (with its options if so specified) if it is in the group; do not list it otherwise. It's not what 0.69.9 does (it errors out if both group and name are specified). It is not a nice implementation, but  if we agree that the test is valid, we can refactor list.py, possibly using stuff from the tool meister to read the on-disk structures into an in-memory structure that can be diced and sliced in different ways. But not for this PR.